File: tools/yaml_to_json_converter.py
#!/usr/bin/env python3
"""
Convert TFF YAML configuration to JSON format for RP2040
"""

import yaml
import json
import sys
import os

def convert_tff_yaml_to_json(yaml_file, json_file):
    """Convert TFF YAML configuration to JSON format"""

    # Read the YAML file
    with open(yaml_file, 'r') as f:
        yaml_data = yaml.safe_load(f)

    # Convert to simplified JSON structure for RP2040
    json_data = {
        "settings": {
            "overlap_threshold_ms": 100
        },
        "mappings": []
    }

    # Process combos
    if 'combos' in yaml_data:
        for combo in yaml_data['combos']:
            if 'keys' in combo and 'outKeys' in combo:
                # Convert keys to array if it's a string
                keys = combo['keys']
                if isinstance(keys, str):
                    keys = keys.split()

                # Convert outKeys to array if it's a string
                out_keys = combo['outKeys']
                if isinstance(out_keys, str):
                    out_keys = [out_keys]

                json_data['mappings'].append({
                    "keys": keys,
                    "outKeys": out_keys
                })

    # Write JSON file
    with open(json_file, 'w') as f:
        json.dump(json_data, f, indent=2)

    print(f"Converted {yaml_file} to {json_file}")
    return True

if __name__ == "__main__":
    if len(sys.argv) != 3:
        print("Usage: python3 yaml_to_json_converter.py <input.yaml> <output.json>")
        sys.exit(1)

    yaml_file = sys.argv[1]
    json_file = sys.argv[2]

    if not os.path.exists(yaml_file):
        print(f"Error: {yaml_file} not found")
        sys.exit(1)

    try:
        convert_tff_yaml_to_json(yaml_file, json_file)
    except Exception as e:
        print(f"Error converting file: {e}")
        sys.exit(1)

File: tools/yaml_to_json_converter.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <sstream>

// Simple function to convert TFF YAML to JSON
bool convertTffYamlToJson(const std::string& yamlFile, const std::string& jsonFile) {
    std::ifstream inFile(yamlFile);
    if (!inFile.is_open()) {
        std::cerr << "Failed to open input file: " << yamlFile << std::endl;
        return false;
    }

    std::ofstream outFile(jsonFile);
    if (!outFile.is_open()) {
        std::cerr << "Failed to create output file: " << jsonFile << std::endl;
        return false;
    }

    // Write JSON header
    outFile << "{\n";
    outFile << "  \"settings\": {\n";
    outFile << "    \"overlap_threshold_ms\": 100\n";
    outFile << "  },\n";
    outFile << "  \"mappings\": [\n";

    std::string line;
    bool inCombosSection = false;
    std::string currentKeys;
    std::string currentOutKeys;
    bool firstMapping = true;

    while (std::getline(inFile, line)) {
        // Skip empty lines and comments
        if (line.empty() || line[0] == '#') {
            continue;
        }

        // Look for combos section
        if (line.find("combos:") != std::string::npos) {
            inCombosSection = true;
            continue;
        }

        // Process combo entries
        if (inCombosSection) {
            if (line.find("- keys:") != std::string::npos) {
                // Extract keys
                size_t startPos = line.find(":") + 2;
                if (startPos < line.length()) {
                    currentKeys = line.substr(startPos);
                    // Trim whitespace
                    currentKeys.erase(0, currentKeys.find_first_not_of(" \t"));
                    currentKeys.erase(currentKeys.find_last_not_of(" \t") + 1);
                }
            } else if (line.find("outKeys:") != std::string::npos) {
                // Extract outKeys
                size_t startPos = line.find(":") + 2;
                if (startPos < line.length()) {
                    currentOutKeys = line.substr(startPos);
                    // Trim whitespace
                    currentOutKeys.erase(0, currentOutKeys.find_first_not_of(" \t"));
                    currentOutKeys.erase(currentOutKeys.find_last_not_of(" \t") + 1);
                }

                // Write the mapping
                if (!currentKeys.empty() && !currentOutKeys.empty()) {
                    if (!firstMapping) {
                        outFile << ",\n";
                    }

                    outFile << "    {\n";
                    outFile << "      \"keys\": \"" << currentKeys << "\",\n";
                    outFile << "      \"outKeys\": \"" << currentOutKeys << "\"\n";
                    outFile << "    }";

                    firstMapping = false;
                    currentKeys.clear();
                    currentOutKeys.clear();
                }
            }
        }
    }

    // Close JSON structure
    outFile << "\n  ]\n";
    outFile << "}\n";

    inFile.close();
    outFile.close();

    std::cout << "Converted " << yamlFile << " to " << jsonFile << std::endl;
    return true;
}

int main(int argc, char* argv[]) {
    if (argc != 3) {
        std::cerr << "Usage: " << argv[0] << " <input.yaml> <output.json>" << std::endl;
        return 1;
    }

    std::string yamlFile = argv[1];
    std::string jsonFile = argv[2];

    if (!convertTffYamlToJson(yamlFile, jsonFile)) {
        return 1;
    }

    return 0;
}